In [76]:
import pandas as pd
import seaborn as sns
import matplotlib as plt

In [77]:
df = pd.read_csv('CreditDefaultIndices.csv')

In [78]:
df.describe()

SPX Index   VIX Index  (L1)  Freddie Mac 30 Year Mortgage Rate  \
count  3522.00000      3535.000000                        3483.000000   
mean   2570.34655        18.502198                           4.312679   
std    1084.80998         7.111229                           1.095748   
min    1022.58000         9.140000                           2.820000   
25%    1684.32500        13.610000                           3.690000   
50%    2333.26500        16.690000                           4.010000   
75%    3348.98000        21.350000                           4.510000   
max    5029.73000        82.690000                           8.090000   

       U.S. 5-year Treasury constant maturity rate  CDX HY Spread  \
count                                  3497.000000    2883.000000   
mean                                      1.796025     385.511771   
std                                       1.016121      80.879694   
min                                       0.190000     268.971000   
25%                                       1.100000     327.316000   
50%                                       1.620000     356.035000   
75%                                       2.270000     434.685500   
max                                       4.950000     871.112000   

       CDX IG Spread  
count    3111.000000  
mean       72.530379  
std        19.033077  
min        43.750000  
25%        59.274000  
50%        67.836000  
75%        81.725500  
max       151.798000

*There is an inflection point at August 8th, 2012. This is when Bloomberg just started recording Daily End Prices for the CDX High Yield Target.*

In [79]:
import datetime as dt

df['Date'] = pd.to_datetime(df['Date'])

df = df[df['Date'] > '2012/08/08']

df.isnull().sum()

Date                                            0
SPX Index                                      20
VIX Index  (L1)                                 7
Freddie Mac 30 Year Mortgage Rate              55
U.S. 5-year Treasury constant maturity rate    41
CDX HY Spread                                  37
CDX IG Spread                                  38
dtype: int64

In [80]:
df_sp_missing = df[df['SPX Index '].isna()]

df_sp_missing.head()

Date  SPX Index   VIX Index  (L1)  \
24  2024-01-15         NaN            13.25   
59  2023-11-23         NaN            12.80   
117 2023-09-04         NaN            13.82   
161 2023-07-04         NaN            13.70   
172 2023-06-19         NaN            14.19   

     Freddie Mac 30 Year Mortgage Rate  \
24                                 NaN   
59                                 NaN   
117                                NaN   
161                                NaN   
172                                NaN   

     U.S. 5-year Treasury constant maturity rate  CDX HY Spread  CDX IG Spread  
24                                           NaN            NaN            NaN  
59                                           NaN            NaN            NaN  
117                                          NaN            NaN            NaN  
161                                          NaN            NaN            NaN  
172                                          NaN            NaN            NaN

*S&P Values are missing on federal holidays (MLK, Presidents, Good Friday, Memorial Day, 4th July, Labor Day, Thanksgiving)*

*Remove any missing samples at these dates*

In [81]:
df = df.drop(df_sp_missing.index)

df.shape

(2899, 7)

*We are left with 2,899 samples. Let's re-explore null values.* 

In [82]:
df.isnull().sum()

Date                                            0
SPX Index                                       0
VIX Index  (L1)                                 0
Freddie Mac 30 Year Mortgage Rate              38
U.S. 5-year Treasury constant maturity rate    22
CDX HY Spread                                  22
CDX IG Spread                                  22
dtype: int64

In [83]:
df_other_missing = df[df['CDX HY Spread'].isnull() | df['CDX IG Spread'].isnull()]

df_other_missing

Date  SPX Index   VIX Index  (L1)  \
92   2023-10-09     4335.66            17.70   
326  2022-11-11     3992.93            22.52   
350  2022-10-10     3612.39            32.45   
582  2021-11-11     4649.27            17.66   
605  2021-10-11     4361.19            20.00   
835  2020-11-11     3572.66            23.45   
857  2020-10-12     3534.22            25.07   
1088 2019-11-11     3087.01            12.69   
1108 2019-10-14     2966.15            14.57   
1339 2018-11-12     2726.22            20.45   
1364 2018-10-08     2884.43            15.69   
1615 2017-10-09     2544.73            10.33   
1842 2016-11-11     2164.45            14.17   
1866 2016-10-10     2163.66            13.38   
2095 2015-11-11     2075.00            16.06   
2117 2015-10-12     2017.46            16.17   
2348 2014-11-11     2039.68            12.92   
2369 2014-10-13     1874.74            24.64   
2600 2013-11-11     1771.89            12.53   
2620 2013-10-14     1710.14            16.07   
2851 2012-11-12     1380.03            16.68   
2876 2012-10-08     1455.88            15.11   

      Freddie Mac 30 Year Mortgage Rate  \
92                                  NaN   
326                                 NaN   
350                                 NaN   
582                                 NaN   
605                                 NaN   
835                                 NaN   
857                                 NaN   
1088                                NaN   
1108                                NaN   
1339                                NaN   
1364                                NaN   
1615                                NaN   
1842                                NaN   
1866                                NaN   
2095                                NaN   
2117                                NaN   
2348                                NaN   
2369                                NaN   
2600                                NaN   
2620                                NaN   
2851                                NaN   
2876                                NaN   

      U.S. 5-year Treasury constant maturity rate  CDX HY Spread  \
92                                            NaN            NaN   
326                                           NaN            NaN   
350                                           NaN            NaN   
582                                           NaN            NaN   
605                                           NaN            NaN   
835                                           NaN            NaN   
857                                           NaN            NaN   
1088                                          NaN            NaN   
1108                                          NaN            NaN   
1339                                          NaN            NaN   
1364                                          NaN            NaN   
1615                                          NaN            NaN   
1842                                          NaN            NaN   
1866                                          NaN            NaN   
2095                                          NaN            NaN   
2117                                          NaN            NaN   
2348                                          NaN            NaN   
2369                                          NaN            NaN   
2600                                          NaN            NaN   
2620                                          NaN            NaN   
2851                                          NaN            NaN   
2876                                          NaN            NaN   

      CDX IG Spread  
92              NaN  
326             NaN  
350             NaN  
582             NaN  
605             NaN  
835             NaN  
857             NaN  
1088            NaN  
1108            NaN  
1339            NaN  
1364            NaN  
1615            NaN  
1842            NaN  
1866            NaN  
2095            NaN  

**There is a bizarre trend in the missing values around October 10th and November 10th each year. We are investigating why these are missing. We suspect there is some event in the Financial Markets / Financial Tech space to explain this as there are missing values at 3 different sources - Freddie Mac, US Treasury, and IHS Markit CDX Spreads**

*Update: Colby eventually theorized correctly these days were Bank Holidays, but not Federal Holidays. He was correct. Banking observes Columbus Day and Veteran's Day annually. Remove these samples*

In [84]:
df = df.drop(df_other_missing.index)

In [85]:
df_mort_missing = df[df['Freddie Mac 30 Year Mortgage Rate'].isnull()]

df_mort_missing

Date  SPX Index   VIX Index  (L1)  \
57   2023-11-27     4550.43            12.69   
86   2023-10-17     4373.20            17.88   
89   2023-10-12     4349.61            16.69   
135  2023-08-09     4467.71            15.96   
150  2023-07-19     4565.72            13.76   
181  2023-06-06     4283.85            13.96   
422  2022-06-30     3785.38            28.71   
478  2022-04-12     4397.45            24.26   
520  2022-02-10     4504.08            23.91   
836  2020-11-10     3545.53            24.80   
1018 2020-02-24     3225.89            25.03   
1100 2019-10-24     3010.29            13.71   
1217 2019-05-09     2870.72            19.10   
1225 2019-04-29     2943.03            13.11   
1375 2018-09-21     2929.67            11.68   
1546 2018-01-18     2798.03            12.22   

      Freddie Mac 30 Year Mortgage Rate  \
57                                  NaN   
86                                  NaN   
89                                  NaN   
135                                 NaN   
150                                 NaN   
181                                 NaN   
422                                 NaN   
478                                 NaN   
520                                 NaN   
836                                 NaN   
1018                                NaN   
1100                                NaN   
1217                                NaN   
1225                                NaN   
1375                                NaN   
1546                                NaN   

      U.S. 5-year Treasury constant maturity rate  CDX HY Spread  \
57                                           4.38        408.103   
86                                           4.86        493.864   
89                                           4.69        489.583   
135                                          4.12        434.206   
150                                          3.98        418.827   
181                                          3.85        447.987   
422                                          3.01        579.087   
478                                          2.66        408.335   
520                                          1.96        360.230   
836                                          0.46        332.289   
1018                                         1.21        326.430   
1100                                         1.58        334.067   
1217                                         2.25        347.398   
1225                                         2.32        326.400   
1375                                         2.95        315.978   
1546                                         2.43        301.806   

      CDX IG Spread  
57           63.791  
86           75.404  
89           74.840  
135          67.357  
150          65.738  
181          71.204  
422         101.159  
478          72.817  
520          65.580  
836          52.976  
1018         53.071  
1100         54.619  
1217         63.053  
1225         57.745  
1375         61.535  
1546         47.992

*There is no real trend for missing Freddie Mac Mortgage Values, so I am going to keep these as there are no other null values.*

In [86]:
df.isnull().sum()

Date                                            0
SPX Index                                       0
VIX Index  (L1)                                 0
Freddie Mac 30 Year Mortgage Rate              16
U.S. 5-year Treasury constant maturity rate     0
CDX HY Spread                                   0
CDX IG Spread                                   0
dtype: int64

*For Developers: Clean Dataset*

In [87]:
df.to_csv('CLEANED_CDX.csv')